Add instructions on how to configure an Azure storage account as a snapshot repository (#6932)
opensearch-project/documentation-website: _tuning-your-cluster/availability-and-recovery/snapshots/snapshot-restore.md

You will most likely not need to specify any parameters except for `bucket` and `base_path`. For allowed request parameters, see [Register or update snapshot repository API](https://opensearch.org/docs/latest/api-reference/snapshots/create-repository/).

### Azure storage account

These steps explain how to register a snapshot repository backed by an Azure storage account for an OpenSearch cluster deployed using Helm.

> 1. Create an Azure storage account. Then create a container within the storage account.

Reviewer: ```suggestion
1. Create an Azure storage account. Then create a container within the storage account. For more information, see the [Azure Storage documentation](https://learn.microsoft.com/en-us/azure/storage/common/storage-introduction).
```

> 1. Create a Docker image with the `repository-azure` plugin.

Author: Hi @ticheng-aws , I have added the command in the Dockerfile below.

> 1. Create a Docker image with the `repository-azure` plugin.

Reviewer 2: @nilushancosta thank you.

> 1. Create a Docker image with the `repository-azure` plugin.

Reviewer: ```suggestion
1. Create an OpenSearch keystore for the Azure storage using the following command:
```

Author: @Naarcha-AWS the user will not be required to execute the bash script below to create the keystore. The user only needs to create the file. Then when the Docker image is built, the script will be copied into the image and executed. Therefore if we say `Create an OpenSearch keystore for the Azure storage using the following command:` it will give the impression that the user needs to create the keystore by executing the script. Instead shall we say `Create the following bash script`?

>    create-keystore.yaml

Reviewer: ```suggestion

```

   ```bash
   #!/bin/bash

   /usr/share/opensearch/bin/opensearch-keystore create
   echo $AZURE_SNAPSHOT_STORAGE_ACCOUNT | /usr/share/opensearch/bin/opensearch-keystore add --stdin azure.client.default.account
   echo $AZURE_SNAPSHOT_STORAGE_ACCOUNT_KEY | /usr/share/opensearch/bin/opensearch-keystore add --stdin azure.client.default.key
   cp /usr/share/opensearch/config/opensearch.keystore /tmp/keystore/opensearch.keystore
   ```

>    Dockerfile

Reviewer: ```suggestion
1. Create a Docker image with the `repository-azure` plugin using the following commands:
```

Author: Since this section has both the Dockerfile and the command to build it and also creates the keystore, shall we say `Create a Docker image with the repository-azure plugin and keystore using the following Dockerfile and command:`

Reviewer: Let's separate these into two different steps. I'll make the adjustments on the file directly and let you know once I'm done.

Reviewer: Done.

   ```docker
   FROM opensearchproject/opensearch:{{site.opensearch_version}}

   RUN /usr/share/opensearch/bin/opensearch-plugin install --batch repository-azure
   COPY --chmod=0775 create-keystore.sh create-keystore.sh
   ```

   ```
   docker build -t opensearch-custom:{{site.opensearch_version}} -f Dockerfile .
   ```

> 1. Create a Kubernetes secret containing the Azure storage account key

Reviewer: ```suggestion
1. Create a Kubernetes secret containing the Azure storage account key using the following command:

```

Author: I have only added the Kubernetes manifest here. Shall add the `kubectl` command as well? And then say `Create a Kubernetes secret containing the Azure storage account key using the following manifest and command:`

Reviewer: That works for me @nilushancosta

   ```yaml
   apiVersion: v1
   kind: Secret
   metadata:
     name: opensearch
   data:
     azure-snapshot-storage-account-key: ### Insert base64 encoded key
   ```

> 1. Deploy OpenSearch using Helm with the following additional values. Specify the value of the storage account in the `AZURE_SNAPSHOT_STORAGE_ACCOUNT` environment variable.

Reviewer: @ticheng-aws: Would the instructions for installing OpenSearch with Helm suffice? https://opensearch.org/docs/latest/install-and-configure/install-opensearch/helm/

> 1. Deploy OpenSearch using Helm with the following additional values. Specify the value of the storage account in the `AZURE_SNAPSHOT_STORAGE_ACCOUNT` environment variable.

Author: @ticheng-aws , do you suggest to add the link which @Naarcha-AWS posted above in the instructions as reference?

> 1. Deploy OpenSearch using Helm with the following additional values. Specify the value of the storage account in the `AZURE_SNAPSHOT_STORAGE_ACCOUNT` environment variable.

Reviewer 2: Hi @Naarcha-AWS and @nilushancosta, yes, I believe including that link will be helpful. Thank you,again.

> 1. Deploy OpenSearch using Helm with the following additional values. Specify the value of the storage account in the `AZURE_SNAPSHOT_STORAGE_ACCOUNT` environment variable.

Author: Added the link through https://github.com/opensearch-project/documentation-website/pull/6932/commits/dd8343a3dd989e4aeb090f34756a2f84ca26d761

> 1. [Deploy OpenSearch using Helm](https://opensearch.org/docs/latest/install-and-configure/install-opensearch/helm/) with the following additional values. Specify the value of the storage account in the `AZURE_SNAPSHOT_STORAGE_ACCOUNT` environment variable.

Reviewer: ```suggestion
1. [Deploy OpenSearch using Helm](https://opensearch.org/docs/latest/install-and-configure/install-opensearch/helm/) with the following additional values. Specify the value of the storage account in the `AZURE_SNAPSHOT_STORAGE_ACCOUNT` environment variable:

```

   ```yaml
   extraInitContainers:
   - name: keystore-generator
     image: opensearch-custom:{{site.opensearch_version}}
     command: ["/bin/bash", "-c"]
     args: ["bash create-keystore.sh"]
     env:
     - name: AZURE_SNAPSHOT_STORAGE_ACCOUNT
       value: ### Insert storage account name
     - name: AZURE_SNAPSHOT_STORAGE_ACCOUNT_KEY
       valueFrom:
         secretKeyRef:
           name: opensearch
           key: azure-snapshot-storage-account-key
     volumeMounts:
     - name: keystore
       mountPath: /tmp/keystore

   extraVolumeMounts:
   - name: keystore
     mountPath: /usr/share/opensearch/config/opensearch.keystore
     subPath: opensearch.keystore
  
   extraVolumes:
   - name: keystore
     emptyDir: {}

   image:
     repository: "opensearch-custom"
     tag: {{site.opensearch_version}}
   ```

> 1. Register the repository using the REST API. Replace `snapshot_container` with the name you specified earlier in step 1.

Reviewer: ```suggestion
1. Register the repository using the Snapshot API. Replace `snapshot_container` with the name you specified earlier in step 1, as shown in the following command:
```

You will most likely not need to specify any parameters except for `bucket` and `base_path`. For allowed request parameters, see [Register or update snapshot repository API](https://opensearch.org/docs/latest/api-reference/snapshots/create-repository/).

### Registering an Azure storage account

Use the following steps to register a snapshot repository backed by an Azure storage account for an OpenSearch cluster deployed using Helm.

> 1. Create an Azure storage account. Then create a container within the storage account. For more information, see the [Azure Storage documentation](https://learn.microsoft.com/en-us/azure/storage/common/storage-introduction).

Reviewer 3: ```suggestion
1. Create an Azure storage account. Then create a container within the storage account. For more information, see [Introduction to Azure Storage](https://learn.microsoft.com/en-us/azure/storage/common/storage-introduction).
```

1. Create a Docker image with the `repository-azure` plugin.

   ```bash
   #!/bin/bash

   /usr/share/opensearch/bin/opensearch-keystore create
   echo $AZURE_SNAPSHOT_STORAGE_ACCOUNT | /usr/share/opensearch/bin/opensearch-keystore add --stdin azure.client.default.account
   echo $AZURE_SNAPSHOT_STORAGE_ACCOUNT_KEY | /usr/share/opensearch/bin/opensearch-keystore add --stdin azure.client.default.key
   cp /usr/share/opensearch/config/opensearch.keystore /tmp/keystore/opensearch.keystore
   ```

> 1. Create a Docker file. This file contains the details of your keystore, the OpenSearch instance, and the Azure repository. To create the file, copy the following example and save it as `Dockerfile`: 

Reviewer 3: Should "a" precede `Dockerfile`?

> 1. Create a Docker file. This file contains the details of your keystore, the OpenSearch instance, and the Azure repository. To create the file, copy the following example and save it as `Dockerfile`: 

Reviewer: ```suggestion
1. Create a Docker file. This file contains the details of your keystore, the OpenSearch instance, and the Azure repository. To create the file, copy the following example and save it as a `Dockerfile`: 
```

   ```docker
   FROM opensearchproject/opensearch:{{site.opensearch_version}}

   RUN /usr/share/opensearch/bin/opensearch-plugin install --batch repository-azure
   COPY --chmod=0775 create-keystore.sh create-keystore.sh
   ```

1. Use the following `docker build` command to build an OpenSearch image from your Dockerfile:

   ```
   docker build -t opensearch-custom:{{site.opensearch_version}} -f Dockerfile .
   ```

> 1. Create a Kubernetes secret containing the Azure storage account key using the following manifest and command:

Reviewer 3: ```suggestion
1. Create a Kubernetes secret containing the Azure storage account key by using the following manifest and command:
```

     name: opensearch
   data:
     azure-snapshot-storage-account-key: ### Insert base64 encoded key
   ```

1. [Deploy OpenSearch using Helm](https://opensearch.org/docs/latest/install-and-configure/install-opensearch/helm/) with the following additional values. Specify the value of the storage account in the `AZURE_SNAPSHOT_STORAGE_ACCOUNT` environment variable:

   ```yaml
   extraInitContainers:
   - name: keystore-generator
     image: opensearch-custom:{{site.opensearch_version}}
     command: ["/bin/bash", "-c"]
     args: ["bash create-keystore.sh"]
     env:
     - name: AZURE_SNAPSHOT_STORAGE_ACCOUNT
       value: ### Insert storage account name
     - name: AZURE_SNAPSHOT_STORAGE_ACCOUNT_KEY
       valueFrom:
         secretKeyRef:
           name: opensearch
           key: azure-snapshot-storage-account-key
     volumeMounts:
     - name: keystore
       mountPath: /tmp/keystore

   extraVolumeMounts:
   - name: keystore
     mountPath: /usr/share/opensearch/config/opensearch.keystore
     subPath: opensearch.keystore
  
   extraVolumes:
   - name: keystore
     emptyDir: {}

   image:
     repository: "opensearch-custom"
     tag: {{site.opensearch_version}}
   ```

> 1. Register the repository using the Snapshot API. Replace `snapshot_container` with the name you specified earlier in step 1, as shown in the following command:

Reviewer 3: ```suggestion
1. Register the repository using the Snapshot API. Replace `snapshot_container` with the name you specified in step 1, as shown in the following command:
```
